Project 4/JCheung-P4-README.md

```diff
@@ -77,34 +106,1457 @@ https://dbpedia.org/sparql?default-graph-uri=http%3A%2F%2Fdbpedia.org&query=PREF
+
+```
+Hint: Use these CSV files which are updated each week:
+
+US National Data: https://raw.githubusercontent.com/nytimes/covid-19-data/master/us.csv
+US State Data: https://raw.githubusercontent.com/nytimes/covid-19-data/master/us-states.csv
+US County Data: https://raw.githubusercontent.com/nytimes/covid-19-data/master/us-counties-recent.csv
+
+Hint 2: SPARQL is designed to query RDF data, not CSV data.
+
+Hint 3: Consider using SPARQL property paths to examine the patterns between properties in the data. Property paths reveal the routes between nodes in a graph.
+```
+
+**Reference Solution:**
+
+```
+Part 1: Convert the CSV data to RDF data. 
+
+To convert the NY Times COVID-19 data CSV file to RDF, you could use a tool like the RDF Mapping Language (RML) and its implementation in the RMLMapper. Here's an example of how you could create an RML mapping file for the CSV file:
+
+@prefix rr: <http://www.w3.org/ns/r2rml#>.
+@prefix rml: <http://semweb.mmlab.be/ns/rml#>.
+@prefix ql: <http://semweb.mmlab.be/ns/ql#>.
+@prefix xsd: <http://www.w3.org/2001/XMLSchema#>.
+@prefix nyt: <http://example.org/nytimes/covid-19-data#>.
+
+# Define the data source
+rr:DataSource
+  rml:source "https://raw.githubusercontent.com/nytimes/covid-19-data/master/us.csv";
+  rml:referenceFormulation ql:CSV.
+
+# Define the logical source
+rr:LogicalSource
+  rr:source rr:DataSource;
+  rr:iterator "us".
+
+# Define the subject map
+rr:SubjectMap
+  rr:template "http://example.org/nytimes/covid-19-data/{date}/{state}/{county}";
+  rr:class nyt:Observation.
+
+# Define the predicate-object maps
+rr:PredicateObjectMap
+  rr:predicate nyt:date;
+  rr:objectMap [ rml:reference "date"; ].
+rr:PredicateObjectMap
+  rr:predicate nyt:state;
+  rr:objectMap [ rml:reference "state"; ].
+rr:PredicateObjectMap
+  rr:predicate nyt:county;
+  rr:objectMap [ rml:reference "county"; ].
+rr:PredicateObjectMap
+  rr:predicate nyt:fips;
+  rr:objectMap [ rml:reference "fips"; ].
+rr:PredicateObjectMap
+  rr:predicate nyt:cases;
+  rr:objectMap [ rml:reference "cases"; rr:datatype xsd:integer; ].
+rr:PredicateObjectMap
+  rr:predicate nyt:deaths;
+  rr:objectMap [ rml:reference "deaths"; rr:datatype xsd:integer; ].
+
+This mapping file assumes that you have downloaded the us.csv file from the NY Times GitHub repository and saved it locally. It defines a logical source for the us table in the CSV file, and a subject map that generates URIs for each observation based on the date, state, and county. The predicate-object maps specify how to map each column in the CSV file to RDF predicates and objects.
+
+
+
+Part 2: Query national data and construct new graph
+
+PREFIX covid: <http://example.org/covid/>
+
+CONSTRUCT {
+  ?statistic covid:totalCases ?totalCasesValue ;
+             covid:totalDeaths ?totalDeathsValue .
+}
```

Comment:
This is a really good question. I noticed that I could have gone dozens of different directions with this problem. The reason I did not calculate the death rate is because I wanted to take the approach where I can find a property path pattern of growth  in terms of cases and deaths. However, your suggestion is another approach a queryist can take. I barely hit the tip of the iceberg with this research question and SPARQL's capacities.